### ISSUE TITLE
null value in column "server_host" in Session Management for Postgres

### ISSUE THREAD

**Issue by stevenlivz** [processone/ejabberd#2262]
> What version of ejabberd are you using?

18.01.47

> What operating system (version) are you using?

MacOS High Sierra

> How did you install ejabberd (source, package, distribution)?

Building from official github source source

./configure --prefix=$HOME/ejabberd --enable-pgsql --enable-new-sql-schema --enable-all --disable-tools --disable-pam

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

I have used the latest SQL scripts which has the new server_host column and built ejabberd using the new sql schema option. I have set the session management to be stored in the Postgres database but when i try to log on i get the following error:

```
2018-02-02 08:39:10.960 GMT [96502] ERROR:  null value in column "server_host" violates not-null constraint
2018-02-02 08:39:10.960 GMT [96502] DETAIL:  Failing row contains (1517560750000003, <0.635.0>, ejabberd@localhost, admin, null, web, , [{ip,{{127,0,0,1},51351}},{conn,c2s_tls},{auth_module,ejabberd_a...).
2018-02-02 08:39:10.960 GMT [96502] STATEMENT:  WITH upsert AS (UPDATE sm SET node=$1, username=$2, resource=$3, priority=$4, info=$5 WHERE usec=$6 AND pid=$7 RETURNING *) INSERT INTO sm(usec, pid, node, username, resource, priority, info) SELECT $8, $9, $10, $11, $12, $13, $14 WHERE NOT EXISTS (SELECT * FROM upsert)
```

This ultimately gives "User session not found" errors in the ejabberd log.

**Comment by digz6666** [comment]
I have same problem with mysql database on ejabberd version 18.01.

It prints following error on log:
```
2018-02-02 15:59:54.995 [error] <0.568.0>@ejabberd_sql:check_error:1090 SQL query 'Q73531467' at {ejabberd_auth_sql,174} failed: "#HY000Field 'server_host' doesn't have a default value"
```

The code line is: https://github.com/processone/ejabberd/blob/983aaac765894d053b4a952083f849381accc242/src/ejabberd_auth_sql.erl#L172

Ejabberd version: 18.01
OS: Ubuntu 16.04.03 64bit
Install type: binary installer, deb package

**Comment by alexeyshch** [comment]
Please make sure that ejabberd is recompiled with `--enable-new-sql-schema` and no old modules are left in the system which were compiled without it.  That query should look like this:
```
WITH upsert AS (UPDATE sm SET node=$1, username=$2, server_host=$3, resource=$4, priority=$5, info=$6 WHERE usec=$7 AND pid=$8 RETURNING *) INSERT INTO sm(usec, pid, node, username, server_host, resource, priority, info) SELECT $9, $10, $11, $12, $13, $14, $15, $16 WHERE NOT EXISTS (SELECT * FROM upsert)
```

**Comment by stevenlivz** [comment]
@alexeyshch Per my original issue post above, i am building from the official github source using exactly that flag as shown in the issue above. Is there anything in addition to this that can cause this? I have even done remake just in case. Output is as follows:

```
2018-02-05 14:19:27.087 GMT [16183] ERROR:  null value in column "server_host" violates not-null constraint
2018-02-05 14:19:27.087 GMT [16183] DETAIL:  Failing row contains (1517840367000003, <0.568.0>, ejabberd@localhost, admin, null, web, , [{ip,{{127,0,0,1},57566}},{conn,c2s_tls},{auth_module,ejabberd_a...).
2018-02-05 14:19:27.087 GMT [16183] STATEMENT:  WITH upsert AS (UPDATE sm SET node=$1, username=$2, resource=$3, priority=$4, info=$5 WHERE usec=$6 AND pid=$7 RETURNING *) INSERT INTO sm(usec, pid, node, username, resource, priority, info) SELECT $8, $9, $10, $11, $12, $13, $14 WHERE NOT EXISTS (SELECT * FROM upsert)
```

**Comment by alexeyshch** [comment]
Please try to connect to the running ejabberd using `ejabberdctl debug` and run the following command:
```
ejabberd_sm_sql:module_info(compile).
```

**Comment by stevenlivz** [comment]
Thanks. This shows the following:

```
[{options,[debug_info,nowarn_deprecated_function,
           {i,"include"},
           {i,"deps/fast_xml/include"},
           {i,"deps/xmpp/include"},
           {i,"deps/p1_utils/include"},
           {d,'STRONG_RAND_BYTES'},
           {d,'RAND_UNIFORM'},
           {d,'GB_SETS_ITERATOR_FROM'},
           {d,'SHORT_NAME_HASH'},
           {src_dirs,[asn1,src]},
           {i,"include"}]},
 {version,"7.1.4"},
 {source,"/Users/livz/dev/code/libs/ejabberd/src/ejabberd_sm_sql.erl"}]
```

**Comment by alexeyshch** [comment]
Hm, it should contain `{d,'NEW_SQL_SCHEMA'}`.  Check also if there is `{new_sql_schema, true}.` line in `vars.config` in ejabberd build dir.

**Comment by stevenlivz** [comment]
I also tried using the sql option for auth and it has a similar issue.

`2018-02-05 14:45:20.616 [error] <0.627.0>@ejabberd_sql:check_error:1090 SQL query 'Q73531467' at {ejabberd_auth_sql,174} failed: [{severity,'ERROR'},{86,<<"ERROR">>},{code,<<"23502">>},{message,<<"null value in column \"server_host\" violates not-null constraint">>},{detail,<<"Failing row contains (admin, null, testpwd, , , 0, 2018-02-05 14:45:20.613753).">>},{115,<<"public">>},{116,<<"users">>},{99,<<"server_host">>},{file,<<"execMain.c">>},{line,2008},{routine,<<"ExecConstraints">>}]`

**Comment by stevenlivz** [comment]
Ah ok- i will look at the vars.config etc and do some digging. Thanks for the tip.

**Comment by zinid** [comment]
Seems like you're running old `ejabberd_sm_sql.beam`. Please double check you installed ejabberd correctly.

**Comment by alexeyshch** [comment]
With that line compilation options must contain `{d,'NEW_SQL_SCHEMA'}`.  Please make sure that beam files are recompiled and loaded.  You can check the path with `code:which(ejabberd_sm_sql).` in `ejabberdctl debug`.  If it's correct, then please check compilation options by compiling ejabberd with `./rebar -vv compile`.

**Comment by stevenlivz** [comment]
Thanks to the tip on NEW_SQL_SCHEMA i stepped backwards in the installation. So, although i had removed the compiled ejabberd, then I used the -B to force remake, and i was using the latest github source, rebar itself had cached some files and was not updating them in the cloned source - i was unaware this to be the case - i thought i was telling everything to rebuild itself :-)

I decided to deleted the compiled output, remove the entire cloned local source and again did a git clone and make and that has fixed it.

Thanks for the help!

steven
